casper/src/main/resources/PoS.rhox

```diff
@@ -531,7 +544,86 @@ new  PoS,
                   }
                }
              }
+            } |
+
+            // Initialize pending rewards state for each validator
+            contract PendingRewards(@"init", @validatorKeys, @return) = {
+              new createStateKeyCh in {
+                @ListOps!("map", initialActive, *createStateKeyCh, return) |
+                contract createStateKeyCh(@key, @ackCh) = {
+                  TreeHashMap!("set", pendingRewardsMap, key, {}, ackCh)
+                }
+              }
+            } |
+
+            // Sum of new pending rewards with existing state
+            contract PendingRewards(@"append", @validatorKey, @newRewards, @return) = {
+              new currentRewardsMaybeCh, currentRewardsCh in {
+                // Get existing pending rewards (per validator)
+                TreeHashMap!("get", pendingRewardsMap, validatorKey, *currentRewardsMaybeCh) |
+                for(@currentRewardsMaybe <- currentRewardsMaybeCh) {
+                  match currentRewardsMaybe {
+                    Nil => currentRewardsCh!({})
+                    val => currentRewardsCh!(val)
+                  } |
+                  for (@currentRewards <- currentRewardsCh) {
+                    new mergeFunc, mergeResultCh in {
+                      // Merge current pending rewards with new pending rewards
+                      MergeMaps!(currentRewards, newRewards, *mergeFunc, *mergeResultCh) |
+                      // Sum values for current and new rewards
+                      contract mergeFunc(@v1, @v2, valMergeRet) = {
+                        valMergeRet!(v1 + v2)
+                      } |
+                      for (@mergeResult <- mergeResultCh) {
+                        // Update pending rewards state
+                        TreeHashMap!("set", pendingRewardsMap, validatorKey, mergeResult, return)
+                      }
+                    }
+                  }
+                }
+              }
+            } |
+
+            // Collect rewards from all validators / reset state to zero
+            contract PendingRewards(@"collect", @validatorKeys, @return) = {
```

Comment:
Will different invocations of this contract ever be running at the same time?  Do we have to worry about concurrent access to the TreeHashMap when thinking about getting and setting?

Reply:
I'm not sure. Maybe it would be better solution to have "modify" function on TreeHashMap to atomically swap value?

Reply:
@Isaac-DeFrain maybe it's enough just to keep separate channels to store these values? I'm not sure if per validator state would be beneficial. Am I missing something, can you give some example?

Reply:
> atomically swap value?

Yes, I think that supplying an `update` method on TreeHashMap would be a good idea.  It will atomically read the current value, apply a function to it, and write the new value.

Reply:
@tgrospic I think you're right. Just having separate channels for those updates should do the trick.

Reply:
> 
> 
> @Isaac-DeFrain maybe it's enough just to keep separate channels to store these values? I'm not sure if per validator state would be beneficial. Am I missing something, can you give some example?

@tgrospic , wouldn't the validators themselves and the network in general be interested in validator state? 
I'm also wondering about the best structure for the TreeHashMap .. perhaps validators as branches and the various kinds of rewards and penalties as subbranches, with the corresponding channels as leaves? Does this make sense? This way the Tree fives the structural and the leaves interact with the dynamic ? @pyrocto ?

Reply:
@9rb I'm not completely sure what is your notion of _validator state_. E.g. [PoS.rho](https://github.com/rchain/rchain/blob/1aa0765/casper/src/main/resources/PoS.rhox#L95-L130) already exposes validators bonds, rewards and active. Is this answers your first question?

Reply:
@9rb pending rewards are now implemented very similarly as you described. Every validator is one branch (first level) in the TreeHashMap and it stores Rholang Map with each validator as a key (you can say again - validators as branches (second level)). Leaves of this inner map are Int values for pending rewards.